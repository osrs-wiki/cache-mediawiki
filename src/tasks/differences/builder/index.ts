import differencesBuilder from "./builder";

export default differencesBuilder;
