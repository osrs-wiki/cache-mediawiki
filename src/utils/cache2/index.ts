export * from "./LazyPromise";
export * from "./types";
