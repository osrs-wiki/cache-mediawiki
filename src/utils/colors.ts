export const jagexHSLtoHex = (value: number) => {
  const hue = (value >> 10) & 0x3f;
  const sat = (value >> 7) & 0x07;
  const lum = value & 0x7f;

  console.log(`H: ${hue}, S: ${sat}, L: ${lum}`);
  return hslToHex(hue, sat, lum);
};

export const hslToHex = (h: number, s: number, l: number) => {
  l /= 100;
  const a = (s * Math.min(l, 1 - l)) / 100;
  const f = (n: number) => {
    const k = (n + h / 30) % 12;
    const color = l - a * Math.max(Math.min(k - 3, 9 - k, 1), -1);
    return Math.round(255 * color)
      .toString(16)
      .padStart(2, "0"); // convert to Hex and prefix "0" if needed
  };
  return `#${f(0)}${f(8)}${f(4)}`;
};
