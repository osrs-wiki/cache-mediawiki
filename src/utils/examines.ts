export const getExamines = async (
  name: string,
  version = "master"
): Promise<{ [key: string]: string }> => {
  const examines: { [key: string]: string } = {};
  const req = await fetch(
    `https://raw.githubusercontent.com/Joshua-F/osrs-examines/${version}/${name}.csv`
  );
  if (!req.ok) {
    return;
  }
  const fetchedExamines = await req.text();
  fetchedExamines.split("\n").forEach((line) => {
    const id = line.split(",")[0];
    const examine = line.substring(line.indexOf(",") + 1).replaceAll('"', "");
    if (id) {
      examines[id] = examine;
    }
  });
  return examines;
};
