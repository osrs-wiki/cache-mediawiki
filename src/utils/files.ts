export const formatFileName = (name: string) => {
  return name
    .replaceAll(":", "-")
    .replaceAll("?", "")
    .replaceAll(/<[^>]*>/g, "");
};
