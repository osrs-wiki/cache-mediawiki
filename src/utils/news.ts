import Parser from "rss-parser";

const NEWS_RSS_FEED =
  "https://secure.runescape.com/m=news/latest_news.rss?oldschool=true";

export const getLatestNewsTitle = async (): Promise<
  | {
      title: string;
      date: string;
    }
  | undefined
> => {
  try {
    const parser = new Parser();
    const feed = await parser.parseURL(NEWS_RSS_FEED);

    const latestNews = feed.items
      .filter((item) => item.categories.includes("Game Updates"))
      .sort(
        (a, b) => new Date(b.pubDate).getTime() - new Date(a.pubDate).getTime()
      )?.[0];

    return latestNews
      ? {
          title: latestNews.title,
          date: new Date(latestNews.pubDate).toISOString(),
        }
      : undefined;
  } catch (error) {
    console.error(error);
  }
};
