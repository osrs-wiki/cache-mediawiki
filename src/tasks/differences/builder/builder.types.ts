import {
  Area,
  ConfigType,
  DBRow,
  Enum,
  IndexType,
  Item,
  NPC,
  Obj,
  Param,
  SpotAnim,
  Sprites,
  Struct,
  Varbit,
  VarPlayer,
} from "../../../utils/cache2";
import { Loadable } from "../../../utils/cache2/Loadable";
import { Difference } from "../differences.types";

export type IndexFeature<T extends Loadable, Name> = {
  name: Name;
  identifiers: (keyof T)[];
  fields: (keyof T)[];
  urls?: IndexURLs;
};

export type IndexURLType = "abex" | "chisel";

export type IndexURLs = { [key in IndexURLType]?: string };

export type IndexFeatures =
  | IndexFeature<Area, "Areas">
  | IndexFeature<DBRow, "Database Rows">
  | IndexFeature<Enum, "Enums">
  | IndexFeature<Item, "Items">
  | IndexFeature<NPC, "Npcs">
  | IndexFeature<Obj, "Objects">
  | IndexFeature<Param, "Params">
  | IndexFeature<SpotAnim, "Spot Anims">
  | IndexFeature<Sprites, "Sprites">
  | IndexFeature<Struct, "Structs">
  | IndexFeature<Varbit, "Varbits">;

export const resultNameMap: { [key in Difference]: string } = {
  added: "New",
  changed: "Diff",
  removed: "Removed",
};

export const indexNameMap: {
  [key in IndexType]?: { [key: number]: IndexFeatures } | IndexFeatures;
} = {
  [IndexType.Configs]: {
    [ConfigType.Area]: {
      name: "Areas",
      identifiers: ["name", "id"],
      fields: ["category", "spriteId"],
      urls: {},
    },
    [ConfigType.Object]: {
      name: "Objects",
      identifiers: ["name", "id"],
      fields: ["actions"],
      urls: {
        chisel: "https://chisel.weirdgloop.org/moid/object_id.html#",
        abex: "https://abextm.github.io/cache2/#/viewer/obj/",
      },
    },
    [ConfigType.Enum]: {
      name: "Enums",
      identifiers: ["id"],
      fields: ["defaultValue", "map"],
      urls: {
        chisel:
          "https://chisel.weirdgloop.org/structs/index.html?type=enums&id=",
        abex: "https://abextm.github.io/cache2/#/viewer/enum/",
      },
    },
    [ConfigType.Npc]: {
      name: "Npcs",
      identifiers: ["name", "id"],
      fields: ["combatLevel", "actions"],
      urls: {
        chisel: "https://chisel.weirdgloop.org/moid/npc_id.html#",
        abex: "https://abextm.github.io/cache2/#/viewer/npc/",
      },
    },
    [ConfigType.Item]: {
      name: "Items",
      identifiers: ["name", "id"],
      fields: [
        "examine",
        "isMembers",
        "isGrandExchangable",
        "isStackable",
        "noteLinkedItem",
        "inventoryActions",
        "placeholderLinkedItem",
        "price",
        "weight",
      ],
      urls: {
        chisel: "https://chisel.weirdgloop.org/moid/item_id.html#",
        abex: "https://abextm.github.io/cache2/#/viewer/item/",
      },
    },
    [ConfigType.Params]: {
      name: "Params",
      identifiers: ["id"],
      fields: ["type", "defaultInt", "defaultString", "isMembers"],
      urls: {
        chisel:
          "https://chisel.weirdgloop.org/structs/index.html?type=enums&id=",
        abex: "https://abextm.github.io/cache2/#/viewer/enum/",
      },
    },
    [ConfigType.Struct]: {
      name: "Structs",
      identifiers: ["id"],
      fields: ["params"],
      urls: {
        chisel:
          "https://chisel.weirdgloop.org/structs/index.html?type=structs&id=",
        abex: "https://abextm.github.io/cache2/#/viewer/struct/",
      },
    },
    [ConfigType.DbRow]: {
      name: "Database Rows",
      identifiers: ["id"],
      fields: ["table", "values"],
      urls: {
        abex: "https://abextm.github.io/cache2/#/viewer/dbrow/",
      },
    },
    [ConfigType.SpotAnim]: {
      name: "Spot Anims",
      identifiers: ["id"],
      fields: ["modelId", "animationId"],
      urls: {},
    },
    [ConfigType.VarBit]: {
      name: "Varbits",
      identifiers: ["id"],
      fields: ["index", "leastSignificantBit", "mostSignificantBit"],
      urls: {
        chisel: "https://chisel.weirdgloop.org/varbs/display?varbit=",
      },
    },
  },
  [IndexType.Sprites]: {
    name: "Sprites",
    identifiers: ["id"],
    fields: ["width", "height"],
    urls: {
      chisel: "",
      abex: "https://abextm.github.io/cache2/#/viewer/sprite/",
    },
  },
};
