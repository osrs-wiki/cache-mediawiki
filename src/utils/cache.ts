import { DiskCacheProvider, FlatCacheProvider } from "@abextm/cache2";
import { readFile } from "fs/promises";

export type CacheSource = "github" | "local";
export type CacheFileType = "flat" | "disk";

export const getCacheProviderLocal = async (
  version: string,
  type: CacheFileType
) => {
  if (type === "disk") {
    return getCacheProviderDiskLocal(version);
  } else if (type === "flat") {
    return getCacheProviderFlatLocal(version);
  }
};

export const getCacheProviderDiskLocal = async (version: string) => {
  return new DiskCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/diskcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderFlatLocal = async (version: string) => {
  return new FlatCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/flatcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderGithub = async (version = "master") => {
  return new FlatCacheProvider({
    async getFile(name) {
      const req = await fetch(
        `https://raw.githubusercontent.com/abextm/osrs-cache/${version}/${name}`
      );
      if (!req.ok) {
        return;
      }
      const ab = await req.arrayBuffer();
      return new Uint8Array(ab);
    },
  });
};
