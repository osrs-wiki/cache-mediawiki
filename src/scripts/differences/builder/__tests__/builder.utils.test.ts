import { jagexHSLtoHex } from "../../../../utils/colors";
import { getFieldDifferencesRow, formatEntryValue } from "../builder.utils";

// Mock the jagexHSLtoHex function
jest.mock("../../../../utils/colors", () => ({
  jagexHSLtoHex: jest.fn((color) => `#mock-${color}`),
}));

describe("differences builder utils", () => {
  describe("getFieldDifferencesRow", () => {
    test("one difference", () => {
      expect(
        getFieldDifferencesRow(
          { test: { oldValue: "test removed", newValue: "test added" } },
          "test"
        )
      ).toMatchSnapshot();
    });

    test("multiple difference", () => {
      expect(
        getFieldDifferencesRow(
          {
            test: {
              oldValue: "test removed first",
              newValue: "test added second",
            },
          },
          "test"
        )
      ).toMatchSnapshot();
    });
  });

  describe("formatEntryValue", () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    test("should return empty string for undefined values", () => {
      expect(formatEntryValue("someField", undefined)).toBe("");
    });

    test("should return empty string for empty objects", () => {
      expect(formatEntryValue("someField", {})).toBe("");
    });

    test("should format color fields with HTML span", () => {
      const colorValue = 12345;
      const result = formatEntryValue("color", colorValue);

      expect(jagexHSLtoHex).toHaveBeenCalledWith(colorValue);
      expect(result).toBe(
        `<span style = "background-color: #mock-${colorValue};color:white">${colorValue}</span>`
      );
    });

    test("should format color fields with HTML span when field has 'color' in name", () => {
      const colorValue = 54321;
      const result = formatEntryValue("textColor", colorValue);

      expect(jagexHSLtoHex).toHaveBeenCalledWith(colorValue);
      expect(result).toBe(
        `<span style = "background-color: #mock-${colorValue};color:white">${colorValue}</span>`
      );
    });

    test("should format arrays by converting to comma-separated string", () => {
      const arrayValue = [1, 2, 3];
      expect(formatEntryValue("items", arrayValue)).toBe("[1, 2, 3]");
    });

    test("should replace null/undefined values in arrays with 'None'", () => {
      const arrayWithNulls = [1, null, 3, undefined];
      expect(formatEntryValue("items", arrayWithNulls)).toBe(
        "[1, None, 3, None]"
      );
    });

    test("should return empty string for empty arrays", () => {
      expect(formatEntryValue("emptyArray", [])).toBe("");
    });

    test("should return value as string for string values", () => {
      expect(formatEntryValue("name", "Test String")).toBe("Test String");
    });

    test("should return value as string for numeric values", () => {
      expect(formatEntryValue("count", 42)).toBe("42");
    });

    test("should format objects by converting to string with custom formatting", () => {
      const objectValue = { key1: "value1", key2: 2 };
      expect(formatEntryValue("config", objectValue)).toBe(
        "{'key1': 'value1', 'key2': 2}"
      );
    });

    test("should handle complex nested objects", () => {
      const complexObject = { a: { b: 1 }, c: [2, 3] };
      expect(formatEntryValue("complex", complexObject)).toBe(
        "{'a': {'b': 1}, 'c': [2, 3]}"
      );
    });
  });
});
