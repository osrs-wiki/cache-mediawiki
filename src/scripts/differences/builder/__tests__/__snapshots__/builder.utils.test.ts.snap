// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`differences builder utils getFieldDifferencesRow - multiple difference 1`] = `
Object {
  "cells": Array [
    Object {
      "content": Array [
        MediaWikiText {
          "children": undefined,
          "styling": undefined,
          "value": "test",
        },
      ],
    },
    Object {
      "content": Array [
        MediaWikiText {
          "children": undefined,
          "styling": undefined,
          "value": "test <span style=\\"color: #ee4231\\">removed</span> <span style=\\"color: #ee4231\\">first</span>",
        },
      ],
    },
    Object {
      "content": Array [
        MediaWikiText {
          "children": undefined,
          "styling": undefined,
          "value": "test <span style=\\"color: #6bc71f\\">added</span> <span style=\\"color: #6bc71f\\">second</span>",
        },
      ],
    },
  ],
  "minimal": true,
}
`;

exports[`differences builder utils getFieldDifferencesRow - one difference 1`] = `
Object {
  "cells": Array [
    Object {
      "content": Array [
        MediaWikiText {
          "children": undefined,
          "styling": undefined,
          "value": "test",
        },
      ],
    },
    Object {
      "content": Array [
        MediaWikiText {
          "children": undefined,
          "styling": undefined,
          "value": "test <span style=\\"color: #ee4231\\">removed</span>",
        },
      ],
    },
    Object {
      "content": Array [
        MediaWikiText {
          "children": undefined,
          "styling": undefined,
          "value": "test <span style=\\"color: #6bc71f\\">added</span>",
        },
      ],
    },
  ],
  "minimal": true,
}
`;
